crates/nu-cli/src/env/directory_specific_environment.rs

```diff
@@ -0,0 +1,194 @@
+                        tag: _,
+                    } = dirval
+                    {
+                        return Some(PathBuf::from(&dir));
+                    }
+                    None
+                })
+                .collect()
+        } else {
+            vec![]
+        };
+        whitelisted_directories.sort();
+
+        DirectorySpecificEnvironment {
+            whitelisted_directories,
+            added_env_vars: IndexMap::new(),
+            overwritten_env_values: IndexMap::new(),
+        }
+    }
+
+    pub fn overwritten_values_to_restore(&mut self) -> Result<IndexMap<String, String>> {
+        let current_dir = std::env::current_dir()?;
+
+        let mut keyvals_to_restore = IndexMap::new();
+        let mut new_overwritten = IndexMap::new();
+
+        for (directory, keyvals) in &self.overwritten_env_values {
+            let mut working_dir = Some(current_dir.as_path());
+
+            let mut re_add_keyvals = true;
+            while let Some(wdir) = working_dir {
+                if wdir == directory.as_path() {
+                    re_add_keyvals = false;
+                    new_overwritten.insert(directory.clone(), keyvals.clone());
+                    break;
+                } else {
+                    working_dir = working_dir //Keep going up in the directory structure with .parent()
+                        .ok_or_else(|| {
+                            Error::new(ErrorKind::NotFound, "Root directory has no parent")
+                        })?
+                        .parent();
+                }
+            }
+            if re_add_keyvals {
+                for (k, v) in keyvals {
+                    keyvals_to_restore.insert(
+                        k.clone(),
+                        v.to_str()
+                            .ok_or_else(|| {
+                                Error::new(
+                                    ErrorKind::Other,
+                                    format!("{:?} is not valid unicode", v),
+                                )
+                            })?
+                            .to_string(),
+                    );
+                }
+            }
+        }
+
+        self.overwritten_env_values = new_overwritten;
+        Ok(keyvals_to_restore)
+    }
+
+    pub fn env_vars_to_add(&mut self) -> Result<IndexMap<String, String>> {
+        let current_dir = std::env::current_dir()?;
+
+        let mut vars_to_add = IndexMap::new();
+        for dir in &self.whitelisted_directories {
+            let mut working_dir = Some(current_dir.as_path());
+
+            //Start in the current directory, then traverse towards the root with working_dir to see if we are in a subdirectory of a valid directory.
+            while let Some(wdir) = working_dir {
+                if wdir == dir.as_path() {
+                    let toml_doc = std::fs::read_to_string(wdir.join(".nu").as_path())?
```

Comment:
We've been using .nu as the script file extension. I wonder if we should call this something like `.nu-env`. Totally open to better names :)

Reply:
I did not know that! I chose `.nu` because I wanted a fairly generic name - I found it conceivable that the file might be used for other things in the future. With a generic name, it's easier to just add whatever nushell-related sections you want should the need arise. (this might be a matter of [https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it](url) and it's very possible that a more specific name is appropriate.)

I agree that if .nu is already used as a script file extension, this should be named something else. If you think that a more specific name is the way to go, I think `.nu-env` (or `.nu_env if more idiomatic?) is a great choice.